src/main/java/org/elasticsearch/plugins/repository/DefaultPluginDownloader.java

```diff
@@ -0,0 +1,381 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.plugins.repository;
+
+import com.google.common.collect.Maps;
+import org.apache.lucene.util.IOUtils;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.ElasticsearchTimeoutException;
+import org.elasticsearch.common.unit.TimeValue;
+
+import javax.net.ssl.*;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URL;
+import java.net.URLConnection;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.security.GeneralSecurityException;
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class DefaultPluginDownloader implements PluginDownloader {
+
+    /**
+     * Directory where downloaded files are stored
+     */
+    private final Path root;
+
+    /**
+     * Counter for downloaded files
+     */
+    private static final AtomicInteger downloadNumber = new AtomicInteger(0);
+
+    /**
+     * Default settings for the downloader
+     */
+    private TimeValue timeout = TimeValue.timeValueMillis(0);
+    private String userAgent = null;
+    private String elasticsearchVersion = null;
```

Comment:
I prefer using `final` for the 3 variables here and check in constructor if variable is `null` then set a default value in constructor.